File: src/infrastructure/persistence/mongodb/mappers/catalog-product.mapper.ts
// src/infrastructure/persistence/mongodb/mappers/catalog-product.mapper.ts

import {
  MakeProductId,
  MakeProductLabel,
  MakeProductDescription,
  MakePrice,
  MakePositiveCm,
  MakeImageUrl,
  type ProductCategory,
  type PriceRange
} from "../../../../domain/pilot"
import type { CatalogProduct, CatalogVariant } from "../../../../domain/catalog"

// ============================================
// MONGODB DOCUMENT TYPE
// ============================================

export interface CatalogProductDocument {
  _id: string
  label: string
  description: string
  category: string
  priceRange: string
  variants: Array<{
    _tag: "StandardVariant" | "CustomVariant"
    size?: "STANDARD" | "LARGE"
    dimensions?: { width: number; length: number }
    price?: number
  }>
  images: {
    front: string
    detail: string
    gallery: string[]
  }
  shopifyUrl?: string
  publishedAt: Date
}

// ============================================
// MAPPER: Domain -> Document
// ============================================

export const catalogToDocument = (product: CatalogProduct): CatalogProductDocument => ({
  _id: product.id,
  label: product.label,
  description: product.description,
  category: product.category,
  priceRange: product.priceRange,
  variants: product.variants.map((v) => ({
    _tag: v._tag,
    ...(v._tag === "StandardVariant"
      ? { size: v.size }
      : {
          dimensions: {
            width: v.dimensions.width,
            length: v.dimensions.length
          },
          price: v.price
        })
  })),
  images: {
    front: product.images.front,
    detail: product.images.detail,
    gallery: [...product.images.gallery]
  },
  ...(product.shopifyUrl ? { shopifyUrl: product.shopifyUrl } : {}),
  publishedAt: product.publishedAt
})

// ============================================
// MAPPER: Document -> Domain
// ============================================

export const catalogFromDocument = (doc: CatalogProductDocument): CatalogProduct => ({
  _tag: "CatalogProduct",
  id: MakeProductId(doc._id),
  label: MakeProductLabel(doc.label),
  description: MakeProductDescription(doc.description),
  category: doc.category as ProductCategory,
  priceRange: doc.priceRange as PriceRange,
  variants: doc.variants.map(mapVariant),
  images: {
    front: MakeImageUrl(doc.images.front),
    detail: MakeImageUrl(doc.images.detail),
    gallery: doc.images.gallery.map((url) => MakeImageUrl(url))
  },
  ...(doc.shopifyUrl ? { shopifyUrl: doc.shopifyUrl } : {}),
  publishedAt: doc.publishedAt
})

// ============================================
// HELPER
// ============================================

const mapVariant = (v: CatalogProductDocument["variants"][number]): CatalogVariant => {
  if (v._tag === "CustomVariant") {
    return {
      _tag: "CustomVariant",
      dimensions: {
        width: MakePositiveCm(v.dimensions!.width),
        length: MakePositiveCm(v.dimensions!.length)
      },
      price: MakePrice(v.price!)
    }
  }
  return {
    _tag: "StandardVariant",
    size: v.size!
  }
}

File: apps/server-java/src/main/java/com/maisonamane/infrastructure/persistence/mongodb/mapper/CatalogProductMapper.java
package com.maisonamane.infrastructure.persistence.mongodb.mapper;

import com.maisonamane.domain.catalog.projection.CatalogImages;
import com.maisonamane.domain.catalog.projection.CatalogProduct;
import com.maisonamane.domain.catalog.projection.CatalogVariant;
import com.maisonamane.domain.catalog.projection.CatalogVariant.CatalogDimensions;
import com.maisonamane.domain.pilot.enums.PriceRange;
import com.maisonamane.domain.pilot.enums.ProductCategory;
import com.maisonamane.domain.pilot.valueobject.ImageUrl;
import com.maisonamane.domain.pilot.valueobject.Price;
import com.maisonamane.domain.pilot.valueobject.ProductDescription;
import com.maisonamane.domain.pilot.valueobject.ProductId;
import com.maisonamane.domain.pilot.valueobject.ProductLabel;
import com.maisonamane.infrastructure.persistence.mongodb.document.CatalogProductDocument;
import com.maisonamane.infrastructure.persistence.mongodb.document.CatalogProductDocument.ImagesDoc;
import com.maisonamane.infrastructure.persistence.mongodb.document.CatalogProductDocument.VariantDoc;
import io.vavr.control.Option;
import java.util.List;
import java.util.Map;
import org.springframework.stereotype.Component;

@Component
public class CatalogProductMapper {

    public CatalogProductDocument toDocument(CatalogProduct product) {
        return new CatalogProductDocument(
                product.id().value(),
                product.label().value(),
                product.description().value(),
                product.category().name(),
                product.priceRange().name(),
                mapVariantsToDoc(product.variants()),
                mapImagesToDoc(product.images()),
                product.shopifyUrl().getOrNull(),
                product.publishedAt());
    }

    public CatalogProduct toDomain(CatalogProductDocument doc) {
        return CatalogProduct.create(
                ProductId.of(doc.id()),
                ProductLabel.of(doc.label()),
                ProductDescription.of(doc.description()),
                ProductCategory.valueOf(doc.category()),
                PriceRange.valueOf(doc.priceRange()),
                mapVariantsToDomain(doc.variants()),
                mapImagesToDomain(doc.images()),
                Option.of(doc.shopifyUrl()),
                doc.publishedAt());
    }

    private List<VariantDoc> mapVariantsToDoc(List<CatalogVariant> variants) {
        return variants.stream().map(this::mapVariantToDoc).toList();
    }

    private VariantDoc mapVariantToDoc(CatalogVariant variant) {
        return switch (variant) {
            case CatalogVariant.StandardVariant std -> new VariantDoc(
                    "StandardVariant", std.size().name(), null, null);
            case CatalogVariant.CustomVariant custom -> new VariantDoc(
                    "CustomVariant",
                    "CUSTOM",
                    Map.of(
                            "width", custom.dimensions().width().value(),
                            "length", custom.dimensions().length().value()),
                    custom.price().valueInCentimes());
        };
    }

    private ImagesDoc mapImagesToDoc(CatalogImages images) {
        return new ImagesDoc(
                images.front().value(),
                images.detail().value(),
                images.gallery().stream().map(ImageUrl::value).toList());
    }

    private List<CatalogVariant> mapVariantsToDomain(List<VariantDoc> variants) {
        return variants.stream().map(this::mapVariantToDomain).toList();
    }

    private CatalogVariant mapVariantToDomain(VariantDoc doc) {
        if ("CustomVariant".equals(doc._tag())) {
            return new CatalogVariant.CustomVariant(
                    CatalogDimensions.of(doc.dimensions().get("width"), doc.dimensions().get("length")),
                    Price.of(doc.price()));
        }
        return new CatalogVariant.StandardVariant(
                com.maisonamane.domain.pilot.enums.Size.valueOf(doc.size()));
    }

    private CatalogImages mapImagesToDomain(ImagesDoc doc) {
        return CatalogImages.of(
                ImageUrl.of(doc.front()),
                ImageUrl.of(doc.detail()),
                doc.gallery().stream().map(ImageUrl::of).toList());
    }
}
